<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create Refflat-Style Annotations from Genomic Data — refflat_create • GTF.tool</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create Refflat-Style Annotations from Genomic Data — refflat_create"><meta property="og:description" content="This function generates a refflat-style data frame from genomic annotations. The refflat format
represents gene structures, including transcripts and exons, for use in genomic analyses.
Data must be pre-loaded using load_annotation() followed by create_GTF_df()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GTF.tool</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create Refflat-Style Annotations from Genomic Data</h1>

    <div class="hidden name"><code>refflat_create.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function generates a refflat-style data frame from genomic annotations. The refflat format
represents gene structures, including transcripts and exons, for use in genomic analyses.
Data must be pre-loaded using load_annotation() followed by create_GTF_df().</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">refflat_create</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  geneName <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"transcript_id"</span>,</span>
<span>  genetic_elements <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRANSCRIPT"</span>, <span class="st">"MRNA"</span>, <span class="st">"CDS"</span>, <span class="st">"GENE"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-input">input<a class="anchor" aria-label="anchor" href="#arg-input"></a></dt>
<dd><p>A data frame containing genomic data with the following columns:
- `chr`: Chromosome identifier
- `start`: Start position of the annotation
- `end`: End position of the annotation
- `strand`: Strand information ('+' or '-')
- `gene_name`: Name of the associated gene
- `gene_id`: Unique identifier for the gene
- `transcript_name`: Name of the associated transcript
- `transcript_id`: Unique identifier for the transcript
- `annotationType`: Type of annotation (e.g., 'EXON', 'TRANSCRIPT')</p></dd>


<dt id="arg-genename">geneName<a class="anchor" aria-label="anchor" href="#arg-genename"></a></dt>
<dd><p>A string specifying the column name representing gene names (default: 'gene_name').</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>A string specifying the column name representing gene IDs (default: 'gene_id').</p></dd>


<dt id="arg-genetic-elements">genetic_elements<a class="anchor" aria-label="anchor" href="#arg-genetic-elements"></a></dt>
<dd><p>Character vector (optional). A vector of genetic element types (e.g., 'CDS', 'GENE', 'MRNA') to include when creating transcripts for a RefFlat file.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame in refflat format with the following columns:
  - `geneName`: Gene name (based on the `geneName` parameter)
  - `name`: Gene ID (based on the `name` parameter)
  - `chrom`: Chromosome identifier
  - `strand`: Strand information
  - `txStart`: Start position of the transcript
  - `txEnd`: End position of the transcript
  - `cdsStart`: Start position of the coding sequence (CDS)
  - `cdsEnd`: End position of the CDS
  - `exonCount`: Number of exons in the transcript
  - `exonStarts`: Comma-separated list of exon start positions
  - `exonEnds`: Comma-separated list of exon end positions</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function processes input genomic data in parallel using multiple CPU cores. It filters and processes
annotations for transcripts and exons, ensuring that overlapping regions are resolved. Transcripts and
exons are matched, and the resulting structure is formatted in the refflat style.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Run the function</span></span></span>
<span class="r-in"><span><span class="va">refflat_data</span> <span class="op">&lt;-</span> <span class="fu">refflat_create</span><span class="op">(</span><span class="va">input</span>, geneName <span class="op">=</span> <span class="st">'gene_name'</span>, name <span class="op">=</span> <span class="st">'gene_id'</span>, genetic_elements <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDS"</span>, <span class="st">'GENE'</span>, <span class="st">'MRNA'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in refflat_create(input, geneName = "gene_name", name = "gene_id",     genetic_elements = c("CDS", "GENE", "MRNA")):</span> unused argument (genetic_elements = c("CDS", "GENE", "MRNA"))</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jakub Kubiś - JBioSystem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

